src/platform/backends/lxd/lxd_mount_handler.cpp

```diff
@@ -0,0 +1,109 @@
+          QString("%1/instances/%2").arg(lxd_socket_url.toString(), lxd_virtual_machine->vm_name.c_str())},
+      // make_uuid is a seed based unique id generator, that makes the device_name reproducible if the seed
+      // (target_path) is the same. If the seeds are different, then the generated ids are likely to be different as
+      // well. 27 (25 + 2(d_)) letters is the maximum device name length that LXD can accept and d_ stands for device
+      // name.
+      device_name{mp::utils::make_uuid(target_path)
+                      .remove("-")
+                      .left(length_of_unique_id_without_prefix)
+                      .prepend("d_")
+                      .toStdString()}
+{
+}
+
+void LXDMountHandler::activate_impl(ServerVariant /**/, std::chrono::milliseconds /**/)
+{
+    const VirtualMachine::State state = vm->current_state();
+    if (state != VirtualMachine::State::off && state != VirtualMachine::State::stopped)
+    {
+        throw mp::NativeMountNeedsStoppedVMException(vm->vm_name);
+    }
+
+    mpl::log(mpl::Level::info, std::string(category),
+             fmt::format("initializing native mount {} => {} in '{}'", source, target, vm->vm_name));
+    lxd_device_add();
+}
+
+void LXDMountHandler::deactivate_impl(bool /*force*/)
+{
+    // TODO: remove the throwing and adjust the unit tests once lxd fix the hot-unmount bug
+    const VirtualMachine::State state = vm->current_state();
+    if (state != VirtualMachine::State::off && state != VirtualMachine::State::stopped)
+    {
+        throw std::runtime_error("Please stop the instance " + vm->vm_name + " before unmount it natively.");
+    }
+
+    mpl::log(mpl::Level::info, std::string(category),
+             fmt::format("Stopping native mount \"{}\" in instance '{}'", target, vm->vm_name));
+    lxd_device_remove();
+}
+
+LXDMountHandler::~LXDMountHandler() = default;
+
+void LXDMountHandler::lxd_device_remove()
+{
+    const QJsonObject instance_info = lxd_request(network_manager, "GET", lxd_instance_endpoint);
+    QJsonObject instance_info_metadata = instance_info["metadata"].toObject();
+    QJsonObject device_list = instance_info_metadata["devices"].toObject();
+
+    device_list.remove(device_name.c_str());
+    instance_info_metadata["devices"] = device_list;
+    // TODO: make this put method If-Match pattern
+    const QJsonObject json_reply = lxd_request(network_manager, "PUT", lxd_instance_endpoint, instance_info_metadata);
+    lxd_wait(network_manager, multipass::lxd_socket_url, json_reply, timeout_milliseconds);
+}
+
+void LXDMountHandler::lxd_device_add()
+{
+    const QJsonObject instance_info = lxd_request(network_manager, "GET", lxd_instance_endpoint);
+    QJsonObject instance_info_metadata = instance_info["metadata"].toObject();
+    QJsonObject device_list = instance_info_metadata["devices"].toObject();
+
+    const std::string abs_target_path = std::filesystem::path{target}.is_relative() ? "/home/ubuntu/" + target : target;
```

Comment:
Ideally this should be tied to the user in `daemon_config.cpp`. I am going to overlook it though, because I am not sure how big a refactor that would require now and it's about time we get this in. If you find that's not that hard and want to deal with it in a separate PR (e.g. just pass it in within the `VMMount`?), that'd still be welcome :wink:

Reply:
Ok, now I can put the pieces together, I sensed that `"ubuntu"` string should be a variable, but I did not know which one that was. 
I think `ssh_username` is accessible within `LXDMountHanlder` class, because `virutal_machine` class has the `ssh_username` method which calls `LXDVirtualMachine::ssh_username` method,

Reply:
Ahh cool, even better. Well spotted! :+1: